Imputing wealth/parent income and Removing parent savings from parent net worth in college enrollment count

storresrod opened the pull request:
Addressing [Issue #1](https://github.com/UI-Research/baby-bonds/issues/1) by imputing wealth/parent income in new script `nlsy_impute`. 

Main question for team discussion in Issue #1: The current imputation uses a limited set of predictor variables, including student age, sex, and a combined race and ethnicity variable. I originally attempted including a larger set of predictor variables, but multiple imputation with the mice package is sensitive to predictor variables with a lot of missingness and to highly correlated predictors. At the same time, it would be good to review literature to see what types of predictors are traditionally included when imputing missing wealth and income, and to ensure we have sufficient predictor variables for imputation to be more accurate.

Addressing [Issue #3](https://github.com/UI-Research/baby-bonds/issues/3) by removing parent savings from college enrollment count. Adapted the `nlsy_lib` script and the `nlsy_get_col_stat_annual_df` function. Utilized newly imputed net worth and subtracted net savings. If needed, this can be replicated in the function which only looks at fall semester enrollment. 

Also, I created a new joint race and ethnicity variable in the `nlsy_lib` script.

dc0sic commented:
Thank you, Sonia! I have a couple of thought regarding predictors used in imputation. 

Predictors should be mostly parents' characteristics and not youth's, except maybe for race and ethnicity, because youth's race captures parents' race. Aside from race, parents' characteristics should include their age and education, and at the time of the interview (i.e., in 1997). NLSY has mother's age at youth's birth (although it's not in our current sample) but I couldn't find father's age. The survey also has parent's highest grade completed. Some other potential predictors are indicators for renting vs. owning a home, and having retirement savings. 

We should also get a better understanding of what parent's income and wealth contain for youth's who don't live with both parents, and include variables that indicate such living arrangements.    

To verify that predictors are significant and to check for collinearity, you could run a regression of the variable being predicted (e.g. wealth) on the set of predictors being considered, while restricting the sample to those with non-missing wealth observations.

We should also think of steps that would help us validate imputation results. For example, plotting the distributions of non-missing and imputed values would be helpful. Also, because we are imputing income and wealth jointly, a scatterplot of these two variables would be useful.

I hope this is helpful. Aside from these thoughts on imputation, I have some questions regarding changes in nlsy_lig.R. I will leave them as inline comments.

dc0sic commented on NLSY/nlsy_lib.R:
```diff
@@ -1,17 +1,23 @@
+set_urbn_defaults(style = "print")
 # Read the raw data
-nlsydf = readRDS(paste0(here::here(), "/NLSY/NLSY-college-finance.rds"))
+nlsydf = read_csv('NLSY/NLSY-college-finance_imp.csv') %>%
```
I would prefer to keep using NLSY-college-finance.rds as the original NLSY dataset. We can have imputations stored in a separate file and merged onto the main file. Also, considering that we are planning to use the imputations only in R, perhaps we can store them in RDS files as reading from them is faster than from CSV files.

dc0sic commented on NLSY/nlsy_lib.R:
```diff
@@ -1,17 +1,23 @@
 #' Returns the NLSY dataframe with basic demographic information
-#' 
-nlsy_get_base_df = function()
+nlsy_get_base_df = function(data)
```
I would like to understand the reason for this change.

dc0sic commented on NLSY/nlsy_lib.R:
```diff
@@ -243,12 +255,22 @@ nlsy_get_student_loans_df = function()
+  return(sloandf)
+}
+
+# Calling functions
```
I would like to understand the reason for this change.

dc0sic reviewed (commented):
I would like to discuss some of the changes in nlsy_lib.R before we merge this PR.

storresrod commented:
Thanks for the comments and support Damir! I have reverted all suggestions in the lib script, and addressed the comments on the imputation script. Updates on this script include:

storresrod commented:
Thanks for the comments and support Damir! I have reverted all suggestions in the lib script, and addressed the comments on the imputation script. Updates on this script include:
- Adding new parent indicators (such as mom age, homeownership, retirement savings)
- Created new dummy variables for relevant categorical variables (like parent education)
- Checked for significance on subset of data with non-missing income and wealth. Excluded father education and dummy for deceased parents.
- Added new multi-collinearity (correlation plots, stepwise regression, and VIF). Did not identify major concerns, so proceeded with selected parent indicators.
- Created new data frames which show original wealth and income, values for original indicators, and the 10 iterations of newly imputed income and wealth
- Added plots comparing imputed and non-imputed plots (for both income and wealth)
- Added scatter plot comparing income and wealth for non-imputed and each iteration of imputation
If this looks closer to what we would like, we can make a decision about how to call these imputed values in future modelling. Look forward to the team's thoughts!

dc0sic commented on NLSY/nlsy_impute.R:
```diff
@@ -0,0 +1,160 @@
+
+# Making dummy variables
+test$hisp <- ifelse(test$hispanic == "Yes", 1, 0)
+test$ret_sav <- ifelse(test$savings == "YES", 1, 0)
```
I think that the retirement savings variable has many missing values. Do we know how the mice package treats this type of predictor and whether we need to do anything special to prepare it for imputation?

judah-axelrod replied:
According to this vignette, when multiple variables are missing information, `mice` using a series of chained equations in which certain values are set randomly and then iteratively filled in: https://cran.r-project.org/web/packages/miceRanger/vignettes/miceAlgorithm.html. Looks like we don't need to do anything special beyond the pre-processing Sonia already did here

judah-axelrod replied:
Looking at the [original paper](https://www.jstatsoft.org/article/view/v045i03), it looks like we have the following options:
1. Run `mice` as is and the vignette above describes how the joint missingness problem is handled.
2. The authors suggesting dropping predictors that have too much missingness to have any useful information. It looks like retirement savings, home ownership, and father's education are all candidates. Intuition: If there are only a small number of observations where pincome *or* retirement savings is present, then omitting retirement savings will only have a small effect on the overall imputation model for pincome.
3. The authors suggest that if we really want to include a predictor in the imputation model but are concerned there is too much missingness as currently constructed, we can first run `mice` to impute *that* variable. My guess is this isn't drastically different than #1, but it could help to improve upon the initial randomization shown in the vignette diagram.

My inclination is to drop father's education (over 3000 missing obs) but then choose option 1 for remaining vars. But we can look into option 3 depending on how much time you think is worth spending on the imputation piece.

dc0sic commented on NLSY/nlsy_impute.R:
```diff
@@ -0,0 +1,160 @@
+test$both_pars <- ifelse(test$both_parents == "Yes", 1, 0)
+test$par_dec <- ifelse(test$par1_deceased == "Yes" | test$par2_deceased == "Yes", 1, 0) # Either parent deceased
+test$mom_educ_hs <- ifelse(test$mom_education == "12TH GRADE" | test$mom_education == "1ST YEAR COLLEGE" | test$mom_education == "2ND YEAR COLLEGE" |  test$mom_education == "3RD YEAR COLLEGE" |  test$mom_education == "4TH YEAR COLLEGE" |  test$mom_education == "5TH YEAR COLLEGE" |  test$mom_education == "6TH YEAR COLLEGE" |  test$mom_education == "7TH YEAR COLLEGE" |  test$mom_education == "8TH Y ...
+test$dad_educ_hs <- ifelse(test$dad_education == "12TH GRADE" | test$dad_education == "1ST YEAR COLLEGE" | test$dad_education == "2ND YEAR COLLEGE" |  test$dad_education == "3RD YEAR COLLEGE" |  test$mom_education == "4TH YEAR COLLEGE" |  test$mom_education == "5TH YEAR COLLEGE" |  test$mom_education == "6TH YEAR COLLEGE" |  test$mom_education == "7TH YEAR COLLEGE" |  test$mom_education == "8TH Y ...
```
A couple of comments on the education dummies:

1. It would be good to have more than two education categories. In addition to "high-school graduates", I would add "some college", "BS degree", "graduate degree".
2. We might want to experiment with different education groupings, which is why it would be good to implement this as a function in nlsy_lib.R. For example, nlsy_encode_educ4 for a function that encodes 4 categories. Having a function is also good because we will want to do this multiple times (for mother, father, and later for the respondent), and we don't want to copy-paste the code.
3. Rather than creating dummies, we can create factor variables, which R, in regressions, converts into dummies. 
4. Rather than comparing a variable to each of the values, a more efficient and less error-prone way to implement this type of encoding is by using the operator %in%.
 
The code that implements all of the suggestions above could look like this:
```
nlsy_encode_educ4 = function(var)
{
    return( case_when(
        {{var}} %in% c(
            "12TH GRADE"
        ) ~ "High-school graduate",
        {{var}} %in% c(
            "1ST YEAR COLLEGE",
            "2ND YEAR COLLEGE",
            "3RD YEAR COLLEGE"
        ) ~ "Some college",
        {{var}} %in% c(
            "4TH YEAR COLLEGE",
            "5TH YEAR COLLEGE"
        ) ~ "College degree",
        {{var}} %in% c(
            "6TH YEAR COLLEGE",
            "7TH YEAR COLLEGE",
            "8TH YEAR COLLEGE"
        ) ~ "Graduate degree"
    )
    )
}

test |>
    mutate(
        mom_educ_cat = factor(nlsy_encode_educ4(mom_education)),
        dad_educ_cat = factor(nlsy_encode_educ4(dad_education))
    )
 ```

judah-axelrod replied:
Seconded on all fronts here!

storresrod replied:
Awesome, thank you for the explanation and the code sample. I have added it to the `nlsy_lib.R` script and called it in the imputation script to generate the factor versions of the education variables.

dc0sic commented on NLSY/nlsy_impute.R:
```diff
@@ -0,0 +1,160 @@
+test$mom_educ_hs <- ifelse(test$mom_education == "12TH GRADE" | test$mom_education == "1ST YEAR COLLEGE" | test$mom_education == "2ND YEAR COLLEGE" |  test$mom_education == "3RD YEAR COLLEGE" |  test$mom_education == "4TH YEAR COLLEGE" |  test$mom_education == "5TH YEAR COLLEGE" |  test$mom_education == "6TH YEAR COLLEGE" |  test$mom_education == "7TH YEAR COLLEGE" |  test$mom_education == "8TH Y ...
+test$dad_educ_hs <- ifelse(test$dad_education == "12TH GRADE" | test$dad_education == "1ST YEAR COLLEGE" | test$dad_education == "2ND YEAR COLLEGE" |  test$dad_education == "3RD YEAR COLLEGE" |  test$mom_education == "4TH YEAR COLLEGE" |  test$mom_education == "5TH YEAR COLLEGE" |  test$mom_education == "6TH YEAR COLLEGE" |  test$mom_education == "7TH YEAR COLLEGE" |  test$mom_education == "8TH Y ...
+test$race_w <- ifelse(test$race == "White", 1, 0)
+test$race_b <- ifelse(test$race == "Black or African American", 1, 0)
```
Similar to education categories, we should use more than just these two. We are not limited by the number of race categories in DYNASIM and sample size of a subgroup here does not play a role. For same reasons, I would leave ethnicity in a separate variable from race. For variables that are relevant, we want them to be as rich as possible.

judah-axelrod replied:
Agreed! I think just using all categories and creating a complete set of dummy variables from the original `race` variable would be good here!

storresrod replied:
Great! Have deleted the race dummy and kept the original race variable as a factor.

storresrod replied:
To close the loop on this - the interaction term and the race term were causing the mice package to report a singularity error (likely multicollinearity, which was difficult to pick up given they are both now factor variables). In the final imputation, I have only used the interaction term.

dc0sic commented on NLSY/nlsy_impute.R:
```diff
@@ -0,0 +1,160 @@
+summary(model <- lm(pincome ~ mom_age + hisp + race_b + mom_educ_hs + dad_educ_hs + ret_sav + hown, data = test)) #Adj R2 = 0.2165
+car::vif(model)
+
+summary(model <- lm(pincome ~ mom_age + hisp + race_b + mom_educ_hs + dad_educ_hs + ret_sav + hown + both_pars, data = test)) #Adj R2 = 0.2178
```
I think that it would be good to add interactions between ethnicity and race, as well as between mom's and dad's education.

judah-axelrod replied:
Just flagging a significant amount of dad's education is missing - if we decide to drop it, then that affects the interactions (see my comment above). Otherwise, totally agree on including the interactions!

storresrod replied:
To close the loop on this - the interaction term and the race term were causing the mice package to report a singularity error (likely multicollinearity, which was difficult to pick up given they are both now factor variables). In the final imputation, I have only used the interaction term. Since father education is dropped due to missingness, that interaction is also not created.

dc0sic commented:
Thank you Sonia! This looks good, but if you are moving the imputation code to a Quarto document, maybe we should wait with the merge. Let me know if you have preferences.

I have substantial comments on two things: missing values in variables used as predictors, and encoding of categorical variables and adding their interactions to the predictor set. I am happy to talk if that would be helpful.

dc0sic reviewed (commented):
Thank you Sonia! This looks good, but if you are moving the imputation code to a Quarto document, maybe we should wait with the merge. Let me know if you have preferences.

I have substantial comments on two things: missing values in variables used as predictors, and encoding of categorical variables and adding their interactions to the predictor set. I am happy to talk if that would be helpful.

judah-axelrod commented on NLSY/nlsy_impute.R:
```diff
@@ -0,0 +1,160 @@
+num <- c("mom_age") # numerical variables
+log <- c("hisp", "race_b", "mom_educ_hs", "dad_educ_hs", "ret_sav", "hown", "both_pars") # binary variables
+
+meth[num] = "norm" # numerical variables
+meth[log] = "logreg" # binary variables
```
I don't think this is doing anything right? Since you're setting them all to "" in line 101 below. And looks like we're using the default "pmm" option for the pincome and pnetworth

storresrod replied:
True - this was part of an older attempt - deleted.

dc0sic commented on NLSY/nlsy_imputation.qmd:
```diff
@@ -0,0 +1,227 @@
+        race_eth = ifelse(hisp == "No" & race == "White", "White NonHispanic", race_eth),
+        race_eth = ifelse(hisp == "Yes" & race == "Black or African American", "Black Hispanic", race_eth),
+        race_eth = ifelse(hisp == "No" & race == "Black or African American", "Black NonHispanic", race_eth),
+        race_eth = ifelse(hisp == "Yes" & race == "American Indian, Eskimo, or Aleut", "NonBW Hispanic", race_eth),
```
In my previous comment about coding race and ethnicity, I suggested using all the race categories present in the sample because it would increase the predictive power of this variable. Knowing that a person is American Indian or Asian likely provides more information about the person's income and wealth than if know only that the person is NonBW.

storresrod replied:
Good flag!! I was still thinking about this from a missingness concern, which is not applicable here. Updated!

dc0sic commented on NLSY/nlsy_imputation.qmd:
```diff
@@ -0,0 +1,227 @@
+    select(-id, -hispanic, -race, -mom_education, -dad_education, -savings, -home, -both_parents, -par1_deceased, -par2_deceased, - mom_age_birth, -pincome, -pnetworth, -par_dec, -mom_educ_hs,-dad_educ_hs, -hisp,-race_eth) %>%
+    drop_na()
+
+M <- cor(testing_cor)
```
Showing correlation table would be helpful.

storresrod replied:
Added the correlation table to be displayed right before the correlation plot

dc0sic commented on NLSY/nlsy_imputation.qmd:
````diff
@@ -0,0 +1,227 @@
+```
+
+### Checking for multi-collinearity - VIF and Successive addition of regressors
+
````
Some notes here would be helpful, especially with respect to vif, which I am not familiar with.

judah-axelrod replied:
- The documentation has more technical information about VIF: https://rdrr.io/cran/car/man/vif.html.
- The Wiki has the formula and steps for calculation: https://en.wikipedia.org/wiki/Variance_inflation_factor. The intuition is that you regress a given variable against all other regressors and calculate the coefficient of determination (1/(1-R^2)).
- For practical purposes, a rule of thumb is that VIF values above 10 are considered indications of multicollinearity, and it looks like in this code chunk, none of the values come close to that, which is good.

storresrod replied:
Thanks Judah!

dc0sic commented on NLSY/nlsy_imputation.qmd:
````diff
@@ -0,0 +1,227 @@
+```
+
+## Checking for missings in the predictor variables
+
````
Same here, some explanation would be helpful. What does this ration mr/(mr+mm) tell us and how do we use it to make a decision about the imputation process? For example, (race,  pincome) is 0.6, and (pincome, race) is 0.987. Does this mean that when pincome is missing, race is also missing in 60% of cases, and when race is missing, pincome is also missing in 98.7% of cases?

judah-axelrod replied:
The interpretation of mr/(mr+mm) is "percentage of usable cases to impute row variable from column variable." So in the example above, a ratio of 0.6 means that for 60% of observations with non-missing race, pincome is also non-missing.

Documentation here for reference: https://www.rdocumentation.org/packages/mice/versions/3.16.0/topics/md.pairs

judah-axelrod replied:
The interpretation of mr/(mr+mm) is "percentage of usable cases to impute row variable from column variable." So in the example above, a ratio of 0.6 means that for 60% of observations with missing race, pincome is non-missing and can thus be used in imputation

Documentation here for reference: https://www.rdocumentation.org/packages/mice/versions/3.16.0/topics/md.pairs

storresrod replied:
It's a non-intuitive measure, but it's the one used in the documentation. As Judah shared, a lower value indicates higher missingness in the predictor for the outcome of interest.

dc0sic commented on NLSY/nlsy_imputation.qmd:
```diff
@@ -0,0 +1,227 @@
+
+# Income and Wealth imputation
+set.seed(25)
+imputed = mice(impute, method=meth, predictorMatrix=predM, m=10)
```
I am not sure why we're doing 10 imputation sets; I don't think we need that many. Five or even just three should enough.

storresrod replied:
Fixed down to 5

dc0sic commented on NLSY/nlsy_imputation.qmd:
````diff
@@ -0,0 +1,227 @@
+```
+
+## Scatter Plots for both parent income and wealth (non-imputed is zero)
+
````
The scatter plots reveal some concerning outliers in the imputed sets. The minimum net-worth in the original set is about -$100,000 and some imputed sets have values that approach -$1,000,000. I am not sure what the best way to deal with this is, but one thing we could do is limit min and max values to those from the original set.

It would also be useful to see whisker plots or summary tables with mean, median, min, max, and perhaps 25th and 75th percentile.

storresrod replied:
Added whisker plots! I think this is a great flag. Unfortunately, it does not seem the mice package has this as a built in functionality so we will have to get created. 

I think it would be pretty simple to do this in post-processing, by taking all values that exceed the min and max and replacing them with the data set min and max. However, this doesn't seem to be what we would ideally want - to bound the min and max before the imputation so that the values are generated within those limits. I think we may need to think about what a custom function, with mice embedded into it.
